contracts/pool/src/methods/liquidate.rs

```diff
@@ -72,179 +62,258 @@ fn do_liquidate(
         } else {
-            (who_s_token_balance, who_compounded_collat)
-        };
+            let amount_to_sub = liq_lp_amount
+                .checked_neg()
+                .ok_or(Error::LiquidateMathError)?;
+            s_token_supply = s_token_supply
+                .checked_sub(liq_lp_amount)
+                .ok_or(Error::LiquidateMathError)?;
 
-    let s_token = STokenClient::new(env, &reserve.s_token_address);
-    let mut s_token_supply = read_token_total_supply(env, &reserve.s_token_address);
-    let debt_token_supply = read_token_total_supply(env, &reserve.debt_token_address);
+            s_token.burn(who, &liq_lp_amount, &liq_comp_amount, liquidator);
+            add_stoken_underlying_balance(env, &s_token.address, amount_to_sub)?;
+        }
 
-    if receive_stoken {
-        let mut liquidator_configurator = UserConfigurator::new(env, liquidator, true);
-        let liquidator_config = liquidator_configurator.user_config()?;
+        user_configurator.withdraw(
+            collat.reserve.get_id(),
+            &collat.asset,
+            collat.lp_balance == liq_lp_amount,
+        )?;
 
-        assert_with_error!(
+        write_token_total_supply(env, &collat.reserve.s_token_address, s_token_supply)?;
+        write_token_balance(
             env,
-            !liquidator_config.is_borrowing(env, reserve.get_id()),
-            Error::MustNotHaveDebt
-        );
+            &s_token.address,
+            who,
+            collat.lp_balance - liq_lp_amount,
+        )?;
 
-        let liquidator_collat_before = read_token_balance(env, &s_token.address, liquidator);
-
-        let liquidator_collat_after = liquidator_collat_before
-            .checked_add(s_token_amount)
-            .ok_or(Error::MathOverflowError)?;
+        recalculate_reserve_data(
+            env,
+            &collat.asset,
+            &collat.reserve,
+            s_token_supply,
+            debt_token_supply,
+        )?;
 
-        s_token.transfer_on_liquidation(who, liquidator, &s_token_amount);
-        write_token_balance(env, &s_token.address, liquidator, liquidator_collat_after)?;
+        total_debt_to_cover_in_base += debt_in_base;
 
-        let use_as_collat = liquidator_collat_before == 0;
+        let npv_after = total_collat_disc_after_in_base
+            .checked_sub(total_debt_after_in_base)
+            .ok_or(Error::LiquidateMathError)?;
 
-        liquidator_configurator
-            .deposit(reserve.get_id(), &asset, use_as_collat)?
-            .write();
-    } else {
-        let amount_to_sub = underlying_amount
-            .checked_neg()
-            .ok_or(Error::MathOverflowError)?;
-        s_token_supply = s_token_supply
-            .checked_sub(s_token_amount)
-            .ok_or(Error::MathOverflowError)?;
-
-        s_token.burn(who, &s_token_amount, &underlying_amount, liquidator);
-        add_stoken_underlying_balance(env, &s_token.address, amount_to_sub)?;
+        if npv_after.is_positive() {
```

Comment:
Do we need here to check is position has initial health instead of just positive npv?